drivers/volume/portworx/portworx.go

```diff
@@ -1666,6 +1295,311 @@ func (p *portworx) UpdateMigratedPersistentVolumeSpec(
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// createGroupCloudSnapFromVolumes creates cloud group snapshots
+func (p *portworx) createGroupCloudSnapFromVolumes(
+	groupSnap *stork_crd.GroupVolumeSnapshot,
+	volNames []string,
+	options map[string]string) (
+	*storkvolume.GroupSnapshotCreateResponse, error) {
+
+	credID := getCredIDFromSnapshot(groupSnap.Spec.Options)
+
+	resp, err := p.volDriver.CloudBackupGroupCreate(&api.CloudBackupGroupCreateRequest{
+		CredentialUUID: credID,
+		VolumeIDs:      volNames})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.Names) == 0 {
+		return nil, fmt.Errorf("group cloudsnapshot request returned 0 tasks")
+	}
+
+	return p.generateStatusReponseFromTaskIDs(resp.Names, credID)
+}
+
+// getGroupCloudSnapStatus fetches the current group cloudsnapshot status by using the task ID in the given
+// volumesnapshot object and returns the updated volumesnapshot object
+func (p *portworx) getGroupCloudSnapStatus(snap *stork_crd.GroupVolumeSnapshot) (
+	*storkvolume.GroupSnapshotCreateResponse, error) {
+
+	if len(snap.Status.VolumeSnapshots) == 0 {
+		return nil, fmt.Errorf("group cloudsnapshot has 0 snapshots in status")
+	}
+
+	credID := getCredIDFromSnapshot(snap.Spec.Options)
+
+	taskIDs := make([]string, 0)
+	for _, snapshotStatus := range snap.Status.VolumeSnapshots {
+		taskIDs = append(taskIDs, snapshotStatus.TaskID)
+	}
+	return p.generateStatusReponseFromTaskIDs(taskIDs, credID)
+}
+
+func (p *portworx) generateStatusReponseFromTaskIDs(taskIDs []string, credID string) (*storkvolume.GroupSnapshotCreateResponse, error) {
+	response := &storkvolume.GroupSnapshotCreateResponse{
+		Snapshots: make([]*stork_crd.VolumeSnapshotStatus, 0),
+	}
+
+	failedTasks := make([]string, 0)
+	doneTasks := make([]string, 0)
+	activeTasks := make([]string, 0)
+	snapIDsPendingRevert := make([]string, 0)
+	for _, taskID := range taskIDs {
+		csStatus := p.getCloudSnapStatus(api.CloudBackupOp, taskID)
+		if csStatus.status == api.CloudBackupStatusFailed {
+			logrus.Errorf(csStatus.msg)
+			failedTasks = append(failedTasks, taskID)
+
+			p.revertPXCloudSnaps(snapIDsPendingRevert, credID)
+			snapIDsPendingRevert = make([]string, 0)
+			continue
+		}
+
+		dataSource := &crdv1.VolumeSnapshotDataSource{
+			PortworxSnapshot: &crdv1.PortworxVolumeSnapshotSource{
+				SnapshotID:          csStatus.cloudSnapID,
+				SnapshotType:        crdv1.PortworxSnapshotTypeCloud,
+				SnapshotCloudCredID: credID,
+			},
+		}
+
+		var conditions []crdv1.VolumeSnapshotCondition
+
+		if csStatus.status == api.CloudBackupStatusDone {
+			conditions = getReadySnapshotConditions()
+			doneTasks = append(doneTasks, taskID)
+			snapIDsPendingRevert = append(snapIDsPendingRevert, csStatus.cloudSnapID)
+		} else if csStatus.status != api.CloudBackupStatusNotStarted {
+			conditions = getPendingSnapshotConditions(csStatus.msg)
+			activeTasks = append(activeTasks, taskID)
```

Comment:
What happens to these active/pending cloudsnaps? They need to be canceled and cleaned up too on failure